test/functional/eval/input_spec.lua

```diff
@@ -1,38 +1,292 @@
 local feed = helpers.feed
+local meths = helpers.meths
 local clear = helpers.clear
+local source = helpers.source
 local command = helpers.command
+local exc_exec = helpers.exc_exec
 
 local screen
 
 before_each(function()
   clear()
   screen = Screen.new(25, 5)
   screen:attach()
+  source([[
+    hi Test ctermfg=Red guifg=Red term=bold
+    function CustomCompl(...)
+      return 'TEST'
+    endfunction
+    function CustomListCompl(...)
+      return ['FOO']
+    endfunction
+  ]])
+  screen:set_default_attr_ids({
+    EOB={bold = true, foreground = Screen.colors.Blue1},
+    T={foreground=Screen.colors.Red},
+  })
 end)
 
 describe('input()', function()
   it('works correctly with multiline prompts', function()
     feed([[:call input("Test\nFoo")<CR>]])
     screen:expect([[
-      {1:~                        }|
-      {1:~                        }|
-      {1:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
       Test                     |
       Foo^                      |
-    ]], {{bold=true, foreground=Screen.colors.Blue}})
+    ]])
   end)
   it('works correctly with multiline prompts and :echohl', function()
-    command('hi Test ctermfg=Red guifg=Red term=bold')
     feed([[:echohl Test | call input("Test\nFoo")<CR>]])
     screen:expect([[
-      {1:~                        }|
-      {1:~                        }|
-      {1:~                        }|
-      {2:Test}                     |
-      {2:Foo}^                      |
-    ]], {{bold=true, foreground=Screen.colors.Blue}, {foreground=Screen.colors.Red}})
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {T:Test}                     |
+      {T:Foo}^                      |
+    ]])
+    command('redraw!')
+    screen:expect([[
+                               |
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {T:Foo}^                      |
+    ]])
+  end)
+  it('works correctly with multiple numeric arguments (many args)', function()
+    command('echohl Test')
+    feed([[:call input(1, 2)<CR>]])
+    screen:expect([[
+                               |
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {T:1}2^                       |
+    ]])
+    feed('<BS>')
+    screen:expect([[
+                               |
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {EOB:~                        }|
+      {T:1}^                        |
+    ]])
+  end)
+  it('works correctly with multiple numeric arguments (dict arg)', function()
+    feed([[:echohl Test | echo input({"prompt": 1, "default": 2, "cancelreturn": 3})<CR>]])
+    command('redraw!')  -- Without this it shows `12` on the line above.
```

Comment:
Ideally this redraw! should also go. What the user sees is what the screen looks like before the redraw... If it is a drawing bug, a stop-gap solution is to add also an expect before the "redraw!` and a TODO note.